README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
Create a D3 Chord Diagram
=========================

The `chorddiag` package allows to create interactive chord diagrams using the JavaScript visualization library D3 (<http://d3js.org>) from within R using the `htmlwidgets` interfacing framework.. Chord diagrams show directed relationships among a group of entities. The chord diagram layout is explained in detail here: <https://github.com/mbostock/d3/wiki/Chord-Layout>.

To quote the explanation found there:

> Consider a hypothetical population of people with different hair colors: black, blonde, brown and red. Each person in this population has a preferred hair color for a dating partner; of the 29,630 (hypothetical) people with black hair, 40% (11,975) prefer partners with the same hair color. This preference is asymmetric: for example, only 10% of people with blonde hair prefer black hair, while 20% of people with black hair prefer blonde hair. A chord diagram visualizes these relationships by drawing quadratic Bézier curves between arcs. The source and target arcs represents two mirrored subsets of the total population, such as the number of people with black hair that prefer blonde hair, and the number of people with blonde hair that prefer black hair.

To create a chord diagram for this example, we need the preferences in matrix format:

``` r
m <- matrix(c(11975,  5871, 8916, 2868,
              1951, 10048, 2060, 6171,
              8010, 16145, 8090, 8045,
              1013,   990,  940, 6907),
            byrow = TRUE,
            nrow = 4, ncol = 4)
haircolors <- c("black", "blonde", "brown", "red")
dimnames(m) <- list(have = haircolors,
                    prefer = haircolors)
m
#>         prefer
#> have     black blonde brown  red
#>   black  11975   5871  8916 2868
#>   blonde  1951  10048  2060 6171
#>   brown   8010  16145  8090 8045
#>   red     1013    990   940 6907
```

Then, the `chorddiag` function creates the chord diagram:

``` r
library(chorddiag)
groupColors <- c("#000000", "#FFDD89", "#957244", "#F26223")
chorddiag(m, groupColors = groupColors, groupnamePadding = 20)
```

![Directional chord diagram for the hair dataset](vignettes/images/chorddiagram-directional-hair.png)

This diagram will be interactive when generated by the function. Interactive here means chord fading and tooltip popups on certain mouse over events. E.g. if the mouse pointer hovers over the chord connecting the "blonde" and "red" groups, a tooltip is displayed giving the numbers for the chord, and all other chords fade away.

![Directional chord diagram for the hair dataset](vignettes/images/chorddiagram-directional-hair-tooltip.png)

The package's code is based on <http://bl.ocks.org/mbostock/4062006>, with modifications for fading behaviour and addition of tooltips.

The default chord diagram type is **directional** but chord diagrams can also be a useful visualization of relationships between two categories of groups, i.e. contingency tables. In this package, this type of chord diagram is called **bipartite** (because there are only chords *between* categories but not *within* categories). Here is an example for the `Titanic` dataset, depicting how many passengers from the different classes and from the crew survived or died when the ship sunk:

![Bipartite chord diagram for the Titanic dataset](vignettes/images/chorddiagram-bipartite-titanic.png)

See the [vignette](vignettes/chorddiagram-vignette.Rmd) for more information.

Updating to the latest version of chorddiag
-------------------------------------------

You can track (and contribute to) development of `chorddiag` at <https://github.com/mattflor/chorddiag>. To install it, run the following command (this requires the `devtools` package):

``` r
devtools::install_github("mattflor/chorddiag")
```

If this fails to install the package vignette, try `devtools::install_github("mattflor/chorddiag", build_vignettes = TRUE)`.


NEWS.md
# chorddiag 0.1.3

* Added a `NEWS.md` file to track changes to the package.
* Finally upgraded d3 to version 4, v4.13.0 to be specific.
* Upgrade d3-tip to version 0.81.
* Removed some obsolete external data reducing the package size considerably.


inst/examples/example.R
library(chorddiag)

# data matrix
m <- matrix(c(11975,  5871, 8916, 2868,
               1951, 10048, 2060, 6171,
               8010, 16145, 8090, 8045,
               1013,   990,  940, 6907),
            byrow = TRUE,
            nrow = 4, ncol = 4)
groupNames <- c("black", "blonde", "brown", "red")
row.names(m) <- groupNames
colnames(m) <- groupNames
m

# default call; group arc length is determined by the row sums; here, this is
# the number of people with different hair colors asked what hair color they
# prefer; this reflects how many people there were with the different
# hair colors
chorddiag(m, showGroupnames = T)


# customization: colors, margin, padding
groupColors <- c("#000000", "#FFDD89", "#957244", "#F26223")
chorddiag(m, groupColors = groupColors, groupnamePadding = 50, margin = 100,
          tooltipGroupConnector = " prefer ")
# by transposing the matrix, group arc length is determined by how ofthen the
# different hair colors were prefered:
chorddiag(t(m), groupColors = groupColors, groupnamePadding = 50, margin = 100,
          tooltipGroupConnector = " prefered by ")
# we can also normalize preference by row sums thus expressing for each hair
# color their preference in percentages; this way, each hair color gets the same
# arc length
chorddiag(100*m/rowSums(m), groupColors = groupColors, groupnamePadding = 30,
          margin = 60, tickInterval = 5,
          tooltipGroupConnector = " prefer ", tooltipUnit = " %",
          precision = 2)

# Bipartite chord diagram with Titanic data
if (requireNamespace("dplyr", quietly = TRUE)) {
    library(dplyr)
    titanic_tbl <- tibble::as_tibble(Titanic)
    titanic_tbl <- titanic_tbl %>%
        mutate(across(where(is.character), as.factor))
    by_class_survival <- titanic_tbl %>%
        group_by(Class, Survived) %>%
        summarise(Count = sum(n)) %>%
        ungroup()
    titanic.mat <- matrix(by_class_survival$Count, nrow = 4, ncol = 2, byrow = TRUE)
    dimnames(titanic.mat ) <- list(Class = levels(titanic_tbl$Class),
                                   Survival = levels(titanic_tbl$Survived))
    print(titanic.mat)
    groupColors <- c("#2171b5", "#6baed6", "#bdd7e7", "#bababa", "#d7191c", "#1a9641")
    chorddiag(titanic.mat, type = "bipartite",
              groupColors = groupColors,
              tickInterval = 50)
}
